### PR TITLE
Simplify paco internals

### PR DESCRIPTION
**Pull request by Lysxia**
This is still WIP, but I'm opening this for feedback since it might be a while before I come back to finish this.

The main motivations are:

1. Reduce code duplication (#12)
2. Make paco not depend on any axiom.

### Axioms

The only axiom used is that `JMeq` implies `eq`, which is used by core tactics such as `pcofix` to rewrite the goal/hypotheses so `paco_acc` can be applied. But there is a simpler way by more carefully folding the goal using `sigT`. As a demonstration, the tactics `uncurry_goal` and `curry_goal` in `paco_sigma.v` do the heavy lifting, moving between

    forall x y z, paco2 gf bot2 (blah x) (bleh y z)

and (using simplified notation)

    forall X : { x & y & z & unit }, paco2 gf bot2 (blah (sigproj1 X)) (bleh (sigprojT2 X) (sigprojT3 X))

From there it should be pretty easy to apply a variant of `paco2_acc` to finish.

I think there was something I needed to prove to complete it which motivated the rest, but I don't remember. Otherwise, maybe this is a small part can be spun into a separate patch.

### Unifying

Paco generates a module `pacon.v` and `gpacon.v` for each arity `n`, but they all really follow from the arity 1 case. The `pacon.v` modules encapsulate this reasonably well, but not perfectly, and `gpacon.v` just don't have such sharing. They contain generated but nontrivial coinduction proofs, which makes them (1) difficult to maintain, and (2) slow to compile.

A more modular solution is this:

- The arity 1 case is proved once (`paco_internal.v` which existed, `gpaco_internal.v` which is new).

- The arity n case is also proved once, parameterized by n (`paco_currying.v`, `gpaco_currying.v`, this requires some auxiliary theory about n-ary relations `paco_sigma.v` and `paco_rel.v` (which may be subject to further reorganization)). In fact since this mirrors the arity 1 case closely, it should be pretty straightforward to automate (`gpaco_currying.v` uses such automation).

- However, for various reasons, that general version is not practical to work with, so we still want the individual `pacon.v` and `gpacon.v` proofs, but they should all follow exactly from the above, i.e., using only the tactic `exact`. This is done for the `pacon.v` part, and only halfway done for the `gpacon.v`.

Furthermore, by tweaking namespacing, using modules instead of prefixes `pacon_`, and using sections and local notations, we can eliminate most of the variability in the `pacon.v` files, with just a little at the top and (maybe) the bottom (see `metasrc/paco.py`).

### TODO

In summary, what remains to be done:

- [ ] Replace `pcofix` and other tactics that use `JMeq`
- [ ] Finish `gpaco` refactoring (complete `gpaco_currying` and fix `metasrc/gpaco.py`)
- [ ] Automate proofs in `paco_currying.v` (in the style of `gpaco_currying.v`, but there are probably more improvements to those tactics).

### CODE REVIEW

**Comment by Lysxia** [comment]
I'll close this because I don't plan to work on it in the near future, and the main motivating issue of axioms has been resolved. I'm available to discuss if there's any more interest in this.
